src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp

```diff
@@ -540,7 +540,13 @@
                                                                             \
   product(uintx, ShenandoahAgingCyclePeriod, 1, EXPERIMENTAL,               \
           "With generational mode, increment the age of objects and"        \
-          "regions each time this many young-gen GC cycles are completed.")
- // end of GC_SHENANDOAH_FLAGS
+          "regions each time this many young-gen GC cycles are completed.") \
+                                                                            \
+  notproduct(bool, ShenandoahEnableCardStats, trueInDebug,                  \
+          "Enable statistics collection related to clean & dirty cards")    \
+                                                                            \
+  notproduct(int, ShenandoahCardStatsLogInterval, 50,                       \
```

Comment:
This isn't really cycles right? It's number of workers that completed a card scan?

Reply:
It's a number of card-scan rounds, independently for either RS scan or Update refs, where a round consists of a RS cycle or Update refs cycle by however many worker threads participate. The logging is oblivious to workers, but uses whatever number was used.

Let me attach an example log in the PR summary to illusrate, but roughly speaking, it's as follows:

(start of remembered set (RS) scan)
(end of remembered set scan)

(og of card stats for this round of RS by worker #1)
...
(log of card stats for this round by RS worker #k1)
...
(start of update refs (UR) scan)

(end of update refs scan)

(log of card stats for this round of UR by worker #1)
...
(log of card stats for this round of UR by worker #k2)
...

Every `ShenandoahCardStatsLogInterval` such rounds, in addition to the per round, per worker stats like we did above, we also produce cumulative statistics across all workers and all rounds to date, but one each for RS and UR.

Let me know if that makes sense, and if "cycles" makes sense in the documentation for what I have called "rounds" above.

Reply:
See https://github.com/openjdk/shenandoah/pull/176#issuecomment-1342840919.

Reply:
Please let me know if there is still any confusion wrt the documentation of `ShenandoahCardStatsLogInterval` or if you'd prefer a rewording. Thanks!

Reply:
Maybe "Log cumulative card stats every so many scans of the remembered set"? "Cycle" is a bit overloaded. If I read this, I would expect to see a log message every 50 GC cycles, but with (probably) two rset scans per GC cycle, it would be closer to every 25 GC cycles.

Reply:
I'll reword along the lines of your suggestion.

For the specific example you gave, we will in fact see one cumulative RS log message every 50 RS scans, and one cumulative UR log message every 50 UR scans, thus roughly one each every 50 GC cycles, if you will. The case when they may not be in lockstep might be if there were full gc's or degenerated cycles that did one (e.g. RS) but skipped the other (e.g. UR -- can this happen?), because we maintain two independent counters one for RS scans and one for UR scans.

Reply:
Okay - thank you. Yes, Shenandoah will skip evacuation (and update references) if it finds a sufficient number of regions with no live objects after final mark.